app/services/title_builder.rb

```diff
@@ -4,38 +4,130 @@ class TitleBuilder
   # @param [Array<Cocina::Models::Title>] titles
   # @returns [String] the title value for Solr
   def self.build(titles)
-    title = titles.first
+    title = primary_title(titles) || first_untyped_title(titles) || titles.first
     if title.value
-      title.value
+      my_title = title.value
     elsif title.structuredValue
-      build_structured(title.structuredValue)
+      my_title = title_from_structured_values(title.structuredValue, non_sorting_char_count(title))
     elsif title.parallelValue
-      build(title.parallelValue)
+      my_title = build(title.parallelValue)
     end
+    remove_trailing_punctuation(my_title.strip) if my_title.present?
   end
 
-  def self.build_structured(titles)
-    title_parts = titles.map do |title|
-      if title.structuredValue
-        [build_structured(title.structuredValue), '']
+  # rubocop:disable Metrics/BlockLength
+  # rubocop:disable Metrics/CyclomaticComplexity
+  # rubocop:disable Metrics/MethodLength
+  # rubocop:disable Metrics/PerceivedComplexity
+  def self.title_from_structured_values(structured_values, non_sorting_char_count)
+    structured_title = ''
+    title_from_part = ''
+    structured_values.each do |structured_value|
```

Comment:
This appears to overwrite `structured_title` every time through this loop and then the method returns whatever one was set last?  Is that right?

Reply:
there is only one ultimate value for each structuredValue -- it is the amalgamation of all the diff pieces.